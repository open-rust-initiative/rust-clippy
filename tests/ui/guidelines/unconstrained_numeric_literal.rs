//@aux-build:../auxiliary/proc_macros.rs

#![feature(lint_reasons)]
#![warn(clippy::unconstrained_numeric_literal)]
#![allow(clippy::let_with_type_underscore, clippy::let_and_return)]

extern crate proc_macros;
use proc_macros::{external, inline_macros};

mod basic_expr {
    fn test() {
        let x = 22;
        //~^ ERROR: type of this numeric variable is unconstrained
        //~| NOTE: `-D clippy::unconstrained-numeric-literal` implied by `-D warnings`
        let x = 22.0;
        //~^ ERROR: type of this numeric variable is unconstrained
        // Has type annotation but it's a wildcard.
        let x: _ = 1;
        //~^ ERROR: type of this numeric variable is unconstrained

        let x = 22_i32;
        let x: [i32; 3] = [1, 2, 3];
        let x: (i32, i32) = if true { (1, 2) } else { (3, 4) };
        let x: u64 = 1;
        const CONST_X: i8 = 1;
    }
}

mod nested_local {
    fn test() {
        let x = {
            let y = 1;
            //~^ ERROR: type of this numeric variable is unconstrained
            1
        };

        const CONST_X: i32 = {
            let y = 1;
            //~^ ERROR: type of this numeric variable is unconstrained
            1
        };
    }
}

mod in_macro {
    use super::*;

    #[inline_macros]
    fn internal() {
        inline!(let x = 22;);
        //~^ ERROR: type of this numeric variable is unconstrained
    }

    fn external() {
        external!(let x = 22;);
    }
}

fn check_expect_suppression() {
    #[expect(clippy::unconstrained_numeric_literal)]
    let x = 21;
}

fn main() {}
