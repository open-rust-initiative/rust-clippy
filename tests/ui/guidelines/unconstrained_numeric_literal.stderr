error: type of this numeric variable is unconstrained
  --> $DIR/unconstrained_numeric_literal.rs:12:13
   |
LL |         let x = 22;
   |             ^
   |
note: unconstrained numeric literals defined here
  --> $DIR/unconstrained_numeric_literal.rs:12:17
   |
LL |         let x = 22;
   |                 ^^
   = note: `-D clippy::unconstrained-numeric-literal` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unconstrained_numeric_literal)]`
help: either add suffix to above numeric literal(s) or label the type explicitly
   |
LL |         let x: i32 = 22;
   |             ~~~~~~

error: type of this numeric variable is unconstrained
  --> $DIR/unconstrained_numeric_literal.rs:15:13
   |
LL |         let x = 22.0;
   |             ^
   |
note: unconstrained numeric literals defined here
  --> $DIR/unconstrained_numeric_literal.rs:15:17
   |
LL |         let x = 22.0;
   |                 ^^^^
help: either add suffix to above numeric literal(s) or label the type explicitly
   |
LL |         let x: f64 = 22.0;
   |             ~~~~~~

error: type of this numeric variable is unconstrained
  --> $DIR/unconstrained_numeric_literal.rs:18:13
   |
LL |         let x: _ = 1;
   |             ^^^^
   |
note: unconstrained numeric literals defined here
  --> $DIR/unconstrained_numeric_literal.rs:18:20
   |
LL |         let x: _ = 1;
   |                    ^
help: either add suffix to above numeric literal(s) or label the type explicitly
   |
LL |         let x: i32 = 1;
   |             ~~~~~~

error: type of this numeric variable is unconstrained
  --> $DIR/unconstrained_numeric_literal.rs:32:17
   |
LL |             let y = 1;
   |                 ^
   |
note: unconstrained numeric literals defined here
  --> $DIR/unconstrained_numeric_literal.rs:32:21
   |
LL |             let y = 1;
   |                     ^
help: either add suffix to above numeric literal(s) or label the type explicitly
   |
LL |             let y: i32 = 1;
   |                 ~~~~~~

error: type of this numeric variable is unconstrained
  --> $DIR/unconstrained_numeric_literal.rs:38:17
   |
LL |             let y = 1;
   |                 ^
   |
note: unconstrained numeric literals defined here
  --> $DIR/unconstrained_numeric_literal.rs:38:21
   |
LL |             let y = 1;
   |                     ^
help: either add suffix to above numeric literal(s) or label the type explicitly
   |
LL |             let y: i32 = 1;
   |                 ~~~~~~

error: type of this numeric variable is unconstrained
  --> $DIR/unconstrained_numeric_literal.rs:50:21
   |
LL |         inline!(let x = 22;);
   |                     ^
   |
note: unconstrained numeric literals defined here
  --> $DIR/unconstrained_numeric_literal.rs:50:25
   |
LL |         inline!(let x = 22;);
   |                         ^^
   = note: this error originates in the macro `__inline_mac_fn_internal` (in Nightly builds, run with -Z macro-backtrace for more info)
help: either add suffix to above numeric literal(s) or label the type explicitly
   |
LL |         inline!(let x: i32 = 22;);
   |                     ~~~~~~

error: aborting due to 6 previous errors

